presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,247 @@
+        initializeDimensions();
+    }
+
+    private void initializeDimensions()
+    {
+        List<Integer> bitLengths = new ArrayList<>(encodingBits);
+        for (int dimensionIndex = maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
+            for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
+                if (bitLengths.get(bitLengthIndex) > 0) {
+                    dimensions[dimensionIndex]++;
+                    bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+                }
+            }
+        }
+    }
+
+    void setReturnType(ReturnType returnType)
+    {
+        this.returnType = returnType;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param ranges the list of minimum and maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    List<ZAddressRange> zOrderCondensedSearchCurve(List<ZValueRange> ranges)
+    {
+        this.ranges = fillInNulls(ranges);
+
+        addressRanges.clear();
+
+        findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength);
+
+        return addressRanges;
+    }
+
+    /**
+     * Searches for and outputs the long addresses within certain ranges in each dimension.
+     *
+     * @param ranges the list of minimum and maximum values of each dimension
+     * @return the list of long addresses within the given value ranges
+     */
+    List<Long> zOrderSearchCurveLongs(List<ZValueRange> ranges)
+    {
+        List<Long> addresses = new ArrayList<>();
+
+        List<ZAddressRange> addressRanges = zOrderCondensedSearchCurve(ranges);
+
+        for (ZAddressRange range : addressRanges) {
+            for (long address = range.getMinAddress(); address <= range.getMaxAddress(); address++) {
+                addresses.add(address);
+            }
+        }
+
+        return addresses;
+    }
+
+    /**
+     * Searches for and outputs the integer addresses within certain ranges in each dimension.
+     *
+     * @param ranges the list of minimum and maximum values of each dimension
+     * @return the list of integer addresses within the given value ranges
+     */
+    List<Integer> zOrderSearchCurveIntegers(List<ZValueRange> ranges)
+    {
+        List<Integer> intAddresses = new ArrayList<>();
+
+        List<ZAddressRange> addressRanges = zOrderCondensedSearchCurve(ranges);
+
+        for (ZAddressRange range : addressRanges) {
+            for (long address = range.getMinAddress(); address <= range.getMaxAddress(); address++) {
+                intAddresses.add((int) address);
+            }
+        }
+
+        return intAddresses;
+    }
```

Comment:
This two functions are exactly the same. Can we deduplicate them?

Reply:
Yes! I just searched and found a way to convert a List<Long> into a List<Integer> in one line.